org.gecko.emf.exporter/src/org/gecko/emf/exporter/AbstractEMFExporter.java

```diff
@@ -0,0 +1,1766 @@
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+import java.util.UUID;
+import java.util.stream.Collectors;
+
+import org.eclipse.emf.ecore.EAnnotation;
+import org.eclipse.emf.ecore.EAttribute;
+import org.eclipse.emf.ecore.EClass;
+import org.eclipse.emf.ecore.EDataType;
+import org.eclipse.emf.ecore.EEnum;
+import org.eclipse.emf.ecore.EEnumLiteral;
+import org.eclipse.emf.ecore.EObject;
+import org.eclipse.emf.ecore.EReference;
+import org.eclipse.emf.ecore.EStructuralFeature;
+import org.eclipse.emf.ecore.resource.Resource;
+import org.eclipse.emf.ecore.util.EcoreUtil;
+import org.gecko.emf.exporter.cells.EMFExportEObjectIDValueCell;
+import org.gecko.emf.exporter.cells.EMFExportEObjectManyReferencesValueCell;
+import org.gecko.emf.exporter.cells.EMFExportEObjectOneReferenceValueCell;
+import org.gecko.emf.exporter.cells.EMFExportMappingMatrixReferenceValueCell;
+import org.gecko.emf.exporter.headers.EMFExportEObjectColumnHeader;
+import org.gecko.emf.exporter.headers.EMFExportEObjectGenericColumnHeader;
+import org.gecko.emf.exporter.headers.EMFExportEObjectIDColumnHeader;
+import org.gecko.emf.exporter.headers.EMFExportEObjectManyReferencesColumnHeader;
+import org.gecko.emf.exporter.headers.EMFExportEObjectOneReferenceColumnHeader;
+import org.gecko.emf.exporter.keys.EMFExportRefMatrixNameIDCompositeKey;
+import org.slf4j.Logger;
+
+import com.google.common.base.Stopwatch;
+import com.google.common.collect.HashBasedTable;
+import com.google.common.collect.Table;
+
+/**
+ * Implementation of format-agnostic methods for exporting EMF resources and
+ * lists of EMF objects to desired format.
+ * 
+ * Format-specific implementations should extend this class and only focus on
+ * transformation of matrices resulting from {@link #exportEObjectsToMatrices}
+ * to arrive at desired format.
+ * 
+ * @author Michal H. Siemaszko
+ */
+public abstract class AbstractEMFExporter implements EMFExporter {
+	protected static final List<String> METADATA_ECLASS_MATRIX_COLUMNS_HEADERS = List.of("Name", "Type", "isMany",
+			"isRequired", "Default value", "Documentation");
+
+	protected static final List<String> METADATA_EENUM_MATRIX_COLUMNS_HEADERS = List.of("Name", "Literal", "Value",
+			"Documentation");
+
+	protected static final String METADATA_DOCUMENTATION_HEADER = "Documentation";
+	protected static final String METADATA_PSEUDOID_DOCUMENTATION = "No default ID present, therefore pseudo-ID was generated for identification purposes";
+
+	protected static final String METADATA_MATRIX_NAME_SUFFIX = "Metadata";
+	protected static final String MAPPING_MATRIX_NAME_SUFFIX = "Mapping";
+
+	protected static final String DOCUMENTATION_GENMODEL_SOURCE = "http://www.eclipse.org/emf/2002/GenModel";
+	protected static final String DOCUMENTATION_GENMODEL_DETAILS = "documentation";
+
+	protected static final String ECORE_PACKAGE_NAME = "ecore";
+
+	protected static final String ID_COLUMN_NAME = "Id";
```

Comment:
After giving the whole id/pseudo id topic some more thought, I'd make the following request:
We stick to your mechanism in using the ids or pseudoIds if no ID is available. We should however add a `_id` header for all Objects, regardless if they have and ID field nor not. This would be in line with what mongoEMF, EMFJson and a few others are doing.

Reply:
As you know, currently, if given `EClass` does not have an ID, this "pseudo ID" mechanism kicks in and such "pseudo ID' is generated. In matrix for that given `EClass`, additional column is created, named `id`, and there those "pseudo IDs" are output. Then, during output of data collected in matrices to specific format (CSV, XLSX, ODS, etc.), that `id` column is transferred. So, this is already done, but named differently (i.e. `id`, not `_id` like you mentioned).
 
In addition, metadata sheet for such `EClass` contains additional entry for such "pseudo ID" field, i.e. "No default ID present, therefore pseudo-ID was generated for identification purposes". See screenshots attached.
 
If I rename this to `_id` (from `id`, like currently), would that satisfy your requirement ? 

![Screenshot from 2023-10-24 17-45-36](https://github.com/geckoprojects-org/org.gecko.emf.utils/assets/9023040/5586faa0-eab0-48c3-bb4c-3ba2b63cad3d)

![Screenshot from 2023-10-24 17-45-44](https://github.com/geckoprojects-org/org.gecko.emf.utils/assets/9023040/8c008dd5-d5a2-4793-aea8-89b4579da168)

Reply:
I know and I do understand that. What I'm getting to is the following:

We should always and for every Object have an _id field. This is populated either by the value of the ID Feature or with your pseudo id.  For the simplified family example this would look as follows:

```
family._id, family.id, name
id1,id1,Simpsons
```

With this, there would be always a defined and known id field that can be used, regardless of the name of the id feature. MongoEMF has the same mechanism and this works great.

Reply:
What I meant is that there already ALWAYS is such field and it's named `id`. It is populated with value of `id` attribute of given EObject, if such is defined, otherwise with pseudo-id. But it's already there and it's named `id`. 

That is why I asked whether I should rename `id` to `_id`, since that functionality is already there. From your answer, I understand I should add ANOTHER column, named `_id`, which will essentially be exactly the same as `id` / contain same exact data. Is that correct?

Reply:
In EMF any Attribute can be an id field. We name them id by convention. Thus if we always have and _id field in the export it would a be clearer that this is a generated field. If we use the family example again and remove the id attribute and declare the name as ID field the whole thing would look as follows:

```
family._id, family.name
Simpsons, Simpsons
```

Might seem as overhead, but I think this is clearer.